### DISCUSSION TITLE
ShowEntire fallback scenario

### DISCUSSION THREAD

**Discussion by adaxer** [QuestPDF/QuestPDF#93] [Ideas]
Let's say I have a set of components (of the same type) for eg food recipes. I would like to have them kept together, so I use ShowEntire for the component. 
But as they have different heights, some will fit on one page, but others might not. In this case the LayoutException is thrown. The normal thought would be to have ShowEntire fall back to not being applied for those components and break them to the next page. 
Is this doable?
Thanks!

**Comment by MarcinZiabek** [comment]
Hello 😁 
Can you please post a code sample (and possibly a screenshot) so I can better understand your use case? 

I see two possibilities right now: 
1) You don't need to use the ShowEntire element and you can just rely on the default paging behaviour of the Stack element.
2) You may want to keep every recipe in one block (without paging) but want to have paging support in-between recipes on the list. In such a case, instead of applying the ShowEntire element to the list, you need to apply it to each element of the list.

**Reply by adaxer** [reply]
Hello Marcin,

sure - here the sample. Just create a console app with the code below. If the table gets too many lines, it won't fit on one page, so the LayoutException is thrown. But I can't say at design time when this happens, for the tables that fit, ShowEntire is desired, for the longer ones, it should fall back to not ShowEntire.

Thanks!

````
using QuestPDF.Drawing;
using QuestPDF.Fluent;
using QuestPDF.Helpers;
using QuestPDF.Infrastructure;
using System;
using System.Diagnostics;
using System.Threading.Tasks;

namespace TryShowEntire
{
    internal class Program
    {
        static void Main(string[] args)
        {
            CreatePdf();
        }
        private static void CreatePdf()
        {
                var doc = new PdfDoc();
                doc.GeneratePdf("test.pdf");
            Process.Start("Explorer.exe", "test.pdf");
        }
    }

    public class PdfDoc : IDocument
    {
        public void Compose(IDocumentContainer container)
        {
            container.Page(page =>
            {
                page.Size(PageSizes.A4);
                page.Margin(10);
                page.Header().Background("#CFC").Text("Header");
                page.Footer().Background("#CCF").Text("Footer");
                page.Content().Element(ComposeContent);
            });
        }

        void ComposeContent(IContainer container)
        {
            container.Stack(stack =>
            {
                stack.Item().Text("Title", TextStyle.Default.Size(14));
                stack.Item().Height(20);
                stack.Item().Component<HeaderedTable>(new HeaderedTable("Only 3", 3));
                stack.Item().Component<HeaderedTable>(new HeaderedTable("Has 8", 8));
                stack.Item().Component<HeaderedTable>(new HeaderedTable("Give it about 30", 20)); // Enter a bigger number here
                stack.Item().Component<HeaderedTable>(new HeaderedTable("Properties 5", 5));
            });
        }

        public DocumentMetadata GetMetadata() => new DocumentMetadata();
    }

    public class HeaderedTable : IComponent
    {
        public HeaderedTable(string title, int size = 10)
        {
            Title = title;
            _size = size;
        }

        public string Title { get; set; }

        private int _size;

        public void Compose(IContainer container)
        {
            container.ShowEntire().Stack(stack =>
            {
                stack.Item().BorderBottom(1).Text(Title, TextStyle.Default.Size(12).Bold());
                for (var i = 0; i < _size; i++)
                {
                    stack.Item().BorderBottom(1).Grid(grid =>
                    {
                        grid.Columns(2);
                        grid.Item(1).Text(Placeholders.Name());
                        grid.Item(1).Text(Placeholders.Sentence());
                    });
                }
                stack.Item().Height(10);
            });
        }
    }
}
````

**Comment by MarcinZiabek** [comment]
I understand now 😁 You are looking for the EnsureSpace element, [please take a look at the documentation](https://www.questpdf.com/documentation/api-reference.html#ensure-space).

**Reply by adaxer** [reply]
I see. So, when I set the Value in EnsureSpace to exactly the height of the "body" then smaller tables than a page size will be kept together, and longer ones will start on a new page and be paged without exception - a very nice approach - thanks very much again :-) !

**Reply by MarcinZiabek** [reply]
If you have a better idea on how to name such an element, please let me know :) Some concepts are simple to describe but surprisingly hard to name 😅

**Reply by adaxer** [reply]
Yeah, agree, and it might even be faster to code it than to name it 🤣. I came across EnsureSpace when trying to solve that, but it somehow did not ring the bell then (probably I confused it with "Leave the space free", but that is nonsense as you can do that in other ways). What I could come up with is "ReserveSpace", but it's not convincingly better than EnsureSpace

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
